yew-functional-macro/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+use proc_macro2::Span;
+use quote::{format_ident, quote};
+use syn::parse::{Parse, ParseStream};
+use syn::{parse_macro_input, Block, FnArg, Ident, Item, Type};
+
+struct FunctionalComponent {
+    body: Box<Block>,
+    props_type: Type,
+    props_name: Ident,
+}
+
+impl Parse for FunctionalComponent {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let parsed: Item = input.parse()?;
+
+        match parsed {
+            Item::Fn(func) => {
+                if !func.sig.generics.params.is_empty() {
+                    // TODO maybe find a way to handle those
+                    return Err(syn::Error::new_spanned(
+                        func.sig.generics,
+                        "functional components cannot contain generics",
+                    ));
+                }
+
+                let inputs = &func.sig.inputs;
+
+                if inputs.len() > 1 {
+                    return Err(syn::Error::new_spanned(
+                        inputs,
+                        "functional components must take only parameter of props",
+                    ));
+                }
+
+                let (props_type, props_name) = if let Some(arg) = inputs.into_iter().next() {
+                    match arg {
+                        FnArg::Typed(arg) => {
+                            let ident = match &*arg.pat {
+                                syn::Pat::Ident(ident) => &ident.ident,
+                                pat => {
+                                    return Err(syn::Error::new_spanned(
+                                        pat,
+                                        "Cannot obtain ident. This should never happen",
+                                    ))
+                                }
+                            };
+
+                            let ty = match &*arg.ty {
+                                Type::Reference(ty) => {
+                                    if ty.lifetime.is_some() {
+                                        return Err(syn::Error::new_spanned(&ty.lifetime, "reference must not have life time"));
+                                    }
+
+                                    if ty.mutability.is_some() {
+                                        return Err(syn::Error::new_spanned(&ty.mutability, "reference must not be mutable"));
+                                    }
+
+                                    &*ty.elem
+                                }
+                                ty => return Err(syn::Error::new_spanned(ty, "invalid argument passed. (hint: type must be props struct by reference)"))
+                            };
+
+                            (ty.clone(), ident.clone())
```

Comment:
That suggestion was when there was only type involved. when there's both type and ident, it can't moved so clone is needed. If you know of a way to not clone type and ident, I'd like to know that.

Reply:
The two values don't overlap so there's no reason why you couldn't move them both out of `arg`. I don't think it makes sense to change this now if we're gonna switch to the transparent `FnArg` approach (where it'll make sense to clone the type to avoid dealing with a self-referential struct) but just for reference:
```rust
// arg is owned because we used `into_iter()`
FnArg::Typed(arg) => {
  // this gives us owned values for `pat` and `ty`
  let PatType { attrs, pat, colon_token: _, ty } = arg;

  // extract owned ident from `pat`
  // extract owned type from `ty`
}
```

Reply:
That's good to know.

I switched out to transparent `FnArg` approach but `ty.elem` still needs to be cloned to get the type but I guess that's not a big deal and file as is.